ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs

```diff
@@ -35,5 +36,58 @@ public static class FSharpExpressionUtil
 
     public static readonly Func<IFSharpExpression, bool> IsSimpleValueExpressionFunc = IsSimpleValueExpression;
     public static readonly Func<IFSharpExpression, bool> IsLiteralExpressionFunc = IsLiteralExpression;
+
+    // TODO: change name
+    public static IFSharpExpression GetOuterMostParentExpression(this IFSharpExpression expression)
+    {
+      var currentExpr = expression;
+      while (true)
+      {
+        if (MatchClauseListOwnerExprNavigator.GetByClauseExpression(currentExpr) is { } matchExpr)
+        {
+          currentExpr = matchExpr as IFSharpExpression;
+          continue;
+        }
+        
+        if (SequentialExprNavigator.GetByExpression(currentExpr) is { } seqExpr &&
+            seqExpr.ExpressionsEnumerable.Last() == currentExpr)
+        {
+          currentExpr = seqExpr;
+          continue;
+        }
+
+        if ((IfThenElseExprNavigator.GetByBranchExpression(currentExpr)) is {ElseExpr: { }} ifExpr)
+        {
+          currentExpr = ifExpr;
+          continue;
+        }
+
+        if (LetOrUseExprNavigator.GetByInExpression(currentExpr) is { } letOrUseExpr)
+        {
+          currentExpr = letOrUseExpr;
+          continue;
+        }
+
+        if (LambdaExprNavigator.GetByExpression(currentExpr) is { } lambdaExpr)
+        {
+          currentExpr = lambdaExpr;
+          continue;
+        }
```

Comment:
Likewise, can we go up from any lambda body?

Reply:
See test `FunctionType 01.fs`.

Reply:
I see, thanks. We should carefully consider possible syntax and FCS types here:
```fsharp
let f: int -> double -> string = Unchecked.defaultof<_>
let g: int -> int -> string = f
```
```fsharp
let h: int * (int -> int) = 1, fun x -> ""
```

Reply:
Both these cases don't work correctly with the current implementation. We should disable going up from lambda expressions for now.